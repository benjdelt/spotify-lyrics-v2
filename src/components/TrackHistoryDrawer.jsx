import React, { useState, useEffect, useRef, Children } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCaretRight, faCaretLeft } from '@fortawesome/free-solid-svg-icons'
import '../styles/TrackHistoryDrawer.css';

function DrawerMenuItem(props) {
  const handleClick = () => {
    props.handleClick();
    props.openHandler();
  }

  return (
    <div key={props.key} className="drawer-option" onClick={handleClick} >
      <span className="drawer-option-text">{props.children}</span>
    </div>
  );
}


function TrackHistoryDrawer(props) {

  const wrapperRef = useRef(null);

  const [open, setOpen] = useState(false);

  useEffect(() => {
    document.addEventListener("click", handleClickOutside, false);
    return () => {
      document.removeEventListener("click", handleClickOutside, false);
    };
  }, [])

  const handleClick = event => {
    event.stopPropagation();
    setOpen(!open);
  };

  const handleClickOutside = event => {
    if (wrapperRef.current && !wrapperRef.current.contains(event.target)) {
      setOpen(false);
    }
  };

  const openHandler = () => {
    setOpen(false);
  };

  const closedIcon = faCaretRight;
  const openIcon = faCaretLeft;

  const styles = {
    display: open ? "block" : "none",
  };

  return (
    <div className="drawer-parent">
      <button onClick={handleClick} className="drawer-label" title="dropdown-arrow">
        {props.label} &nbsp;
        <span className="menu-text"><FontAwesomeIcon icon={open ? openIcon : closedIcon} /></span>
      </button>
      <div className="drawer-menu" style={styles} ref={wrapperRef}>
        {Children.map(props.children, child => React.cloneElement(child, { openHandler }))}
      </div>
    </div>
  )
}

export { TrackHistoryDrawer, DrawerMenuItem };
