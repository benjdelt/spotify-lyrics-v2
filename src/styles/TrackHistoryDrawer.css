.drawer-parent {
  height: 2em;
}

.drawer-parent:hover, .drawer-parent:focus {
  text-decoration: none;
  color: #b3b3b3;
}

.drawer-label:hover {
  color: #fff;
}

.drawer-label {
  display: flex;
  align-items: center;
}

.drawer-menu {
  background-color: #1c1c1c;
  margin: .2em;
  padding: .5em; 
  position: relative;
  left: 15em;
  top: -8.5em;
  height: 100vh;
  width: max-content;
  overflow-y: auto;
}

.drawer-option {
  padding: .2em;
  cursor: pointer;
  font-size: .8em;
}

.drawer-option-text {
  display: flex;
  align-items: center;
  padding: .1em;
}

.drawer-option:hover, .drawer-option:focus {
  background-color: #3e3e3e;
  color: #fff;
}

.drawer-option img {
  width: 1.2em;
}

@media (max-width: 1024px) {
  .drawer-menu {
    left: 14em;
  }
}

@media (max-width: 860px) {
  .drawer-label {
    display: initial;
    align-items: initial;
  }

  .drawer-menu {
    display: block;
    position: fixed;
    left: 0em;
    top: 0em;
    height: 87vh;
  }

  .drawer-option-text {
    font-size: .75rem;
  }
}

@media (max-width: 732px) {
  .drawer-menu {
    top: 3em;
    height: 80vh;
  }

  .drawer-option-text {
    font-size: .85rem;
  }
}