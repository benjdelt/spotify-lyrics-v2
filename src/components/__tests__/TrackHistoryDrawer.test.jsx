import { render, cleanup } from '@testing-library/react';
import { TrackHistoryDrawer, DrawerMenuItem } from '../TrackHistoryDrawer';

afterEach(cleanup)

it('should take a snapshot', () => {
  const { asFragment } = render(<DrawerMenuItem />)
  expect(asFragment(<DrawerMenuItem />)).toMatchSnapshot()
});

it('should take a snapshot', () => {
  const { asFragment } = render(<TrackHistoryDrawer />)
  expect(asFragment(<TrackHistoryDrawer />)).toMatchSnapshot()
});
