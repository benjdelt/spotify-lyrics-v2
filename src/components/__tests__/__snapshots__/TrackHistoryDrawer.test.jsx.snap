// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="drawer-option"
  >
    <span
      class="drawer-option-text"
    />
  </div>
</DocumentFragment>
`;

exports[`should take a snapshot 2`] = `
<DocumentFragment>
  <div
    class="drawer-parent"
  >
    <button
      class="drawer-label"
    >
       Â 
      <span
        class="menu-text"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-right fa-w-6 "
          data-icon="caret-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 192 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
            fill="currentColor"
          />
        </svg>
      </span>
    </button>
    <div
      class="drawer-menu"
      style="display: none;"
    />
  </div>
</DocumentFragment>
`;
